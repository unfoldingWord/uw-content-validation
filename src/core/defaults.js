export const DEFAULT_EXCERPT_LENGTH = 20;

export const REPO_CODES_LIST = [
  'UHB', 'UGNT',
  'LT', 'ST',
  'TA', 'TW', 'TWL',
  // The ones with 2 suffix are the 2021 new TSV format repos
  'TN', 'TN2', 'TQ', 'TQ2',
  'SN', 'SQ',
  'UGL', 'UHAL',
  'OBS', 'OBS-TWL',
  'OBS-TN', 'OBS-TN2', 'OBS-TQ', 'OBS-TQ2',
  'OBS-SN', 'OBS-SQ',
];
