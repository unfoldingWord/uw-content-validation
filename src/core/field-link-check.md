## Basic Link Field Check Sandbox

This function is for checking text fields that are links, or that contain links.

```js
// The control code in this box is editable for changing settings—
//    simply click inside here and add, change, or delete text as required.
// Note that (gray) lines starting with // are "comments", i.e., they are ignored by the software
//    so if you want to enable those lines, you must remove the // from the beginning of the line.

import { checkFieldLinks } from './field-link-check';
import { processNoticesToErrorsWarnings } from '../demos/notice-processing-functions';
import { RenderRawResults, RenderSuccessesErrorsWarnings } from '../demos/RenderProcessedResults';
import { userLog } from './utilities';

// Empty, space, link, RC, good, and bad text samples
const textE = "";
const textS = " ";
const textL = "https://unfoldingWord.org";
const textRC1 = "rc://en/obs/book/obs/02/03";
const textRC2 = "rc://en/tw/dict/bible/other/canaan";
const textMD1 = "Have a look [here](https://unfoldingWord.org/)!";
const textMD2 = " Look at [[https://unfoldingWord.org]] ";
const textMD3 = "Now  look at [[http://door43.org]] ";

// Easy for user to change text used in next line (to demonstrate differing results)
const username = 'unfoldingWord';
const chosenTextName = "textMD3";
const chosenText = textMD3;

let rawResults;

// Define our callback function
function acceptUpdatedResult(newLinkResult){
    // Update the results with later results
    userLog(`acceptUpdatedResult callback function is updating result now with ${newLinkResult.noticeList.length} notices.`);
    rawResults.noticeList = newLinkResult.noticeList;
    // Now how can we tell Styleguidist to refresh???
}

const linkOptions = {
    expectedCount: 1, // 0 = all links optional (no minimal number expected)
    allowedCount: 0, // 0 = no maximum limit
    otherTextAllowed: true, // if false, field is just the link only
    linkTypesAllowed: ['RC', 'md', 'naked',],
        // 'RC' (ResourceContainer) allows 'rc://...` links
        // 'md' allows '[this](http://here.org)' and [[http://here.org]]
        // 'naked' allows 'http:://example.org' and even 'example.org'
    checkTargets: true, // Attempt to retrieve the link(s) -- slows it down
    callbackFunction: acceptUpdatedResult, // Required to accept updated results if checking targets
};

// This function returns the results of the fast checks
//  and if specified in linkOptions, the callback will update result later with results of slower checks
rawResults = checkFieldLinks(username, 'en', '', chosenTextName, chosenText, linkOptions, 'that was supplied');
if (!rawResults.successList || !rawResults.successList.length)
    rawResults.successList = ["Done basic link checks"];

<>
<b>Check</b> "{chosenText}"<br/><br/>
<RenderRawResults results={rawResults} />
</>
```
