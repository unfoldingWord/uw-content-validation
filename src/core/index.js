export * from './books/';
export * from './defaults';
export * from './utilities';
export * from './getApi';
export * from './text-handling-functions';

export * from './field-text-check';
export * from './file-text-check';
export * from './lexicon-file-contents-check';
export * from './markdown-text-check';
export * from './markdown-file-contents-check';
export * from './manifest-text-check';
export * from './notes-tsv7-row-check';
export * from './notes-tsv7-table-check';
export * from './plain-text-check';
export * from './questions-tsv7-row-check';
export * from './questions-tsv7-table-check';
export * from './notes-links-check';
export * from './tn-tsv9-row-check';
export * from './tn-tsv9-table-check';
export * from './twl-tsv6-row-check';
export * from './twl-tsv6-table-check';
export * from './usfm-text-check';
export * from './yaml-text-check';
