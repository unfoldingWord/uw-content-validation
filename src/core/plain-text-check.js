// eslint-disable-next-line no-unused-vars
import { DEFAULT_EXCERPT_LENGTH, REPO_CODES_LIST } from './defaults'
import { OPEN_CLOSE_PUNCTUATION_PAIRS, PAIRED_PUNCTUATION_OPENERS, PAIRED_PUNCTUATION_CLOSERS, isWhitespace, countOccurrencesInString } from './text-handling-functions'
import { checkTextField } from './field-text-check';
import { removeDisabledNotices } from './disabled-notices';
// eslint-disable-next-line no-unused-vars
import { parameterAssert } from './utilities';


const PLAIN_TEXT_VALIDATOR_VERSION_STRING = '0.4.0';


/**
 *
 * @param {string} textType 'markdown', 'USFM', 'YAML', 'text', or 'raw'
 * @param {string} repoCode -- e.g., 'TN' or 'TQ2', etc.
 * @param {string} textName
 * @param {string} plainText -- text to be checked
 * @param {string} givenLocation
 * @param {Object} checkingOptions
 */
export function checkPlainText(languageCode, repoCode, textType, textName, plainText, givenLocation, checkingOptions) {
    /* This function is optimised for checking the entire text, i.e., all lines.
        It is used in checkFileContents() in book-package-check.js

    TODO: Should languageCode also be a parameter here? (affects other programs using the API)

     Returns a result object containing a successList and a noticeList
     */
    // functionLog(`checkPlainText(${textName}, (${plainText.length} chars), ${givenLocation}, ${JSON.stringify(checkingOptions)})…`);
    //parameterAssert(languageCode !== undefined, "checkPlainText: 'languageCode' parameter should be defined");
    //parameterAssert(typeof languageCode === 'string', `checkPlainText: 'languageCode' parameter should be a string not a '${typeof languageCode}': ${languageCode}`);
    //parameterAssert(languageCode !== 'markdown' && languageCode !== 'USFM' && languageCode !== 'YAML' && languageCode !== 'text' && languageCode !== 'raw' && languageCode !== 'unfoldingWord', `checkPlainText: 'languageCode' ${languageCode} parameter should be not be '${languageCode}'`);
    //parameterAssert(repoCode !== undefined, "checkPlainText: 'repoCode' parameter should be defined");
    //parameterAssert(typeof repoCode === 'string', `checkPlainText: 'repoCode' parameter should be a string not a '${typeof repoCode}': ${repoCode}`);
    //parameterAssert(REPO_CODES_LIST.includes(repoCode), `checkPlainText: 'repoCode' parameter should not be '${repoCode}'`);
    //parameterAssert(textType !== undefined, "checkPlainText: 'textType' parameter should be defined");
    //parameterAssert(typeof textType === 'string', `checkPlainText: 'textType' parameter should be a string not a '${typeof textType}': ${textType}`);
    //parameterAssert(textType === 'markdown' || textType === 'USFM' || textType === 'YAML' || textType === 'text' || textType === 'raw', `checkPlainText: unrecognised 'textType' parameter: '${textType}'`);
    //parameterAssert(textName !== undefined, "checkPlainText: 'textName' parameter should be defined");
    //parameterAssert(typeof textName === 'string', `checkPlainText: 'textName' parameter should be a string not a '${typeof textName}': ${textName}`);
    //parameterAssert(plainText !== undefined, "checkPlainText: 'plainText' parameter should be defined");
    //parameterAssert(typeof plainText === 'string', `checkPlainText: 'plainText' parameter should be a string not a '${typeof plainText}': ${plainText}`);
    //parameterAssert(checkingOptions !== undefined, "checkPlainText: 'checkingOptions' parameter should be defined");

    let ourLocation = givenLocation;
    if (ourLocation && ourLocation[0] !== ' ') ourLocation = ` ${ourLocation}`;

    let excerptLength;
    try {
        excerptLength = checkingOptions?.excerptLength;
    } catch (ptcError) { }
    if (typeof excerptLength !== 'number' || isNaN(excerptLength)) {
        excerptLength = DEFAULT_EXCERPT_LENGTH;
        // debugLog(`Using default excerptLength=${excerptLength}`);
    }
    // else
    // debugLog(`Using supplied excerptLength=${excerptLength}`, `cf. default=${DEFAULT_EXCERPT_LENGTH}`);
    const excerptHalfLength = Math.floor(excerptLength / 2); // rounded down
    const excerptHalfLengthPlus = Math.floor((excerptLength + 1) / 2); // rounded up
    // debugLog(`Using excerptHalfLength=${excerptHalfLength}`, `excerptHalfLengthPlus=${excerptHalfLengthPlus}`);

    const cptResult = { successList: [], noticeList: [] };

    function addSuccessMessage(successString) {
        // functionLog(`checkPlainText success: ${successString}`);
        cptResult.successList.push(successString);
    }
    function addNotice(noticeObject) {
        // bookID is a three-character UPPERCASE USFM book identifier or 'OBS'.
        // functionLog(`checkPlainText notice: (priority=${priority}) ${message}${characterIndex > 0 ? ` (at character ${characterIndex})` : ""}${excerpt ? ` ${excerpt}` : ""}${location}`);
        //parameterAssert(noticeObject.priority !== undefined, "cPT addNotice: 'priority' parameter should be defined");
        //parameterAssert(typeof noticeObject.priority === 'number', `cPT addNotice: 'priority' parameter should be a number not a '${typeof noticeObject.priority}': ${noticeObject.priority}`);
        //parameterAssert(noticeObject.message !== undefined, "cPT addNotice: 'message' parameter should be defined");
        //parameterAssert(typeof noticeObject.message === 'string', `cPT addNotice: 'message' parameter should be a string not a '${typeof noticeObject.message}': ${noticeObject.message}`);
        if (noticeObject.characterIndex) { //parameterAssert(typeof noticeObject.characterIndex === 'number', `cPT addNotice: 'characterIndex' parameter should be a number not a '${typeof noticeObject.characterIndex}': ${noticeObject.characterIndex}`);
        }
        // //parameterAssert(excerpt!==undefined, "cPT addNotice: 'excerpt' parameter should be defined");
        if (noticeObject.excerpt) { //parameterAssert(typeof noticeObject.excerpt === 'string', `cPT addNotice: 'excerpt' parameter should be a string not a '${typeof noticeObject.excerpt}': ${noticeObject.excerpt}`);
        }
        //parameterAssert(noticeObject.location !== undefined, "cPT addNotice: 'location' parameter should be defined");
        //parameterAssert(typeof noticeObject.location === 'string', `cPT addNotice: 'location' parameter should be a string not a '${typeof noticeObject.location}': ${noticeObject.location}`);

        // noticeObject.debugChain = noticeObject.debugChain ? `checkPlainText(${textType}, ${textName}) ${noticeObject.debugChain}` : `checkPlainText(${textType}, ${textName})`;
        cptResult.noticeList.push(noticeObject);
    }

    function ourCheckTextField(lineNumber, fieldText, allowedLinks, optionalFieldLocation, checkingOptions) {
        /**
        * @description - checks the given text field and processes the returned results
        * @param {string} fieldName - name of the field being checked
        * @param {string} fieldText - the actual text of the field being checked
        * @param {boolean} allowedLinks - true if links are allowed in the field, otherwise false
        * @param {string} optionalFieldLocation - description of where the field is located
        * @param {Object} checkingOptions - parameters that might affect the check
        */
        // Does basic checks for small errors like leading/trailing spaces, etc.

        // We assume that checking for compulsory fields is done elsewhere

        // Updates the global list of notices
        // debugLog(`cPT ourCheckTextField(${fieldName}, (${fieldText.length}), ${allowedLinks}, ${fieldLocation}, …)`);
        //parameterAssert(lineNumber !== undefined, "cPT ourCheckTextField: 'lineNumber' parameter should be defined");
        //parameterAssert(typeof lineNumber === 'number', `cPT ourCheckTextField: 'fieldName' parameter should be a number not a '${typeof lineNumber}'`);
        //parameterAssert(fieldText !== undefined, "cPT ourCheckTextField: 'fieldText' parameter should be defined");
        //parameterAssert(typeof fieldText === 'string', `cPT ourCheckTextField: 'fieldText' parameter should be a string not a '${typeof fieldText}'`);
        //parameterAssert(allowedLinks === true || allowedLinks === false, "cPT ourCheckTextField: allowedLinks parameter must be either true or false");
        //parameterAssert(optionalFieldLocation !== undefined, "cPT ourCheckTextField: 'optionalFieldLocation' parameter should be defined");
        //parameterAssert(typeof optionalFieldLocation === 'string', `cPT ourCheckTextField: 'optionalFieldLocation' parameter should be a string not a '${typeof optionalFieldLocation}'`);

        const resultObject = checkTextField(languageCode, repoCode, textType, '', fieldText, allowedLinks, optionalFieldLocation, checkingOptions);

        // Choose only ONE of the following
        // This is the fast way of append the results from this field
        // cptResult.noticeList = cptResult.noticeList.concat(resultObject.noticeList);
        // If we need to put everything through addNotice, e.g., for debugging or filtering
        //  process results line by line
        for (const noticeEntry of resultObject.noticeList)
            addNotice({ ...noticeEntry, lineNumber });
    }
    // end of ourCheckTextField function

    function checkPlainLineContents(lineNumber, lineText, lineLocation) {

        // functionLog(`checkPlainLineContents for '${lineName}', '${lineText}' at${lineLocation}`);
        let thisText = lineText.trimStart(); // So we don’t get "leading space" AND "doubled spaces" errors

        if (thisText)
            // Allow links as that’s more general
            ourCheckTextField(lineNumber, thisText, true, lineLocation, checkingOptions);
    }
    // end of checkPlainLineContents function


    // Main code for checkPlainText function
    if (isWhitespace(plainText)) {
        addNotice({ priority: 638, message: "Only found whitespace", location: ourLocation });
        return cptResult;
    }

    let characterIndex;
    if ((characterIndex = plainText.indexOf('<<<<<<<')) >= 0) {
        const iy = characterIndex + excerptHalfLength; // Want excerpt to focus more on what follows
        const excerpt = (iy > excerptHalfLength ? '…' : '') + plainText.substring(iy - excerptHalfLength, iy + excerptHalfLengthPlus).replace(/ /g, '␣') + (iy + excerptHalfLengthPlus < plainText.length ? '…' : '')
        addNotice({ priority: 993, message: "Unresolved GIT conflict", characterIndex, excerpt, location: ourLocation });
    } else if ((characterIndex = plainText.indexOf('=======')) >= 0) {
        const iy = characterIndex + excerptHalfLength; // Want excerpt to focus more on what follows
        const excerpt = (iy > excerptHalfLength ? '…' : '') + plainText.substring(iy - excerptHalfLength, iy + excerptHalfLengthPlus).replace(/ /g, '␣') + (iy + excerptHalfLengthPlus < plainText.length ? '…' : '')
        addNotice({ priority: 992, message: "Unresolved GIT conflict", characterIndex, excerpt, location: ourLocation });
    } else if ((characterIndex = plainText.indexOf('>>>>>>>>')) >= 0) {
        const iy = characterIndex + excerptHalfLength; // Want excerpt to focus more on what follows
        const excerpt = (iy > excerptHalfLength ? '…' : '') + plainText.substring(iy - excerptHalfLength, iy + excerptHalfLengthPlus).replace(/ /g, '␣') + (iy + excerptHalfLengthPlus < plainText.length ? '…' : '')
        addNotice({ priority: 991, message: "Unresolved GIT conflict", characterIndex, excerpt, location: ourLocation });
    }

    if (plainText[0] === '\n') {
        characterIndex = 0;
        const excerpt = (plainText.length > excerptLength ? '…' : '') + plainText.slice(-excerptLength).replace(/ /g, '␣').replace(/\n/g, '\\n')
        addNotice({ priority: 539, message: "File starts with empty line", characterIndex, excerpt, location: ourLocation });
    }
    if (!plainText.endsWith('\n') && !textName.endsWith('title.md')) {
        characterIndex = plainText.length - 1;
        const excerpt = (plainText.length > excerptLength ? '…' : '') + plainText.slice(-excerptLength).replace(/ /g, '␣').replace(/\n/g, '\\n')
        addNotice({ priority: 538, message: "File ends without newline character", characterIndex, excerpt, location: ourLocation });
    }
    else if (plainText.endsWith('\n\n')) {
        characterIndex = plainText.length - 2;
        const excerpt = (plainText.length > excerptLength ? '…' : '') + plainText.slice(-excerptLength).replace(/ /g, '␣').replace(/\n/g, '\\n')
        addNotice({ priority: 138, message: "File ends with additional blank line(s)", characterIndex, excerpt, location: ourLocation });
    }

    const lines = plainText.split('\n');
    // debugLog(`  '${location}' has ${lines.length.toLocaleString()} total lines`);
    //  checking nested markers (so that we can give the line number in the notice)
    // let headerLevel = 0;
    // let lastNumLeadingSpaces = 0;
    // let lastLineContents;
    // While checking individual lines,
    const openMarkers = [];
    for (let n = 1; n <= lines.length; n++) {

        const line = lines[n - 1];
        if (line) {
            if (textType === 'text' || textType === 'raw') // other file-types do these checks themselves
                checkPlainLineContents(n, line, ourLocation);

            // Check for nested brackets and quotes, etc.
            for (let characterIndex = 0; characterIndex < line.length; characterIndex++) {
                const char = line[characterIndex];
                let closeCharacterIndex;
                if (PAIRED_PUNCTUATION_OPENERS.indexOf(char) >= 0) {
                    // debugLog(`Saving ${openMarkers.length} '${char}' ${n} ${x}`);
                    openMarkers.push({ char, n, x: characterIndex });
                } else if ((closeCharacterIndex = PAIRED_PUNCTUATION_CLOSERS.indexOf(char)) >= 0) {
                    // debugLog(`Found '${char}' ${n} ${x}`);
                    // debugLog(`Which: ${which} '${openers.charAt(which)}'`)
                    if (openMarkers.length) {
                        const [lastEntry] = openMarkers.slice(-1);
                        // debugLog(`  Recovered lastEntry=${JSON.stringify(lastEntry)}`);
                        // debugLog(`  Comparing found '${char}' with (${which}) '${openers.charAt(which)}' from '${lastEntry.char}'`);
                        if (lastEntry.char === PAIRED_PUNCTUATION_OPENERS.charAt(closeCharacterIndex)) {
                            // debugLog(`  Matched '${char}' with  '${openers.charAt(which)}' ${n} ${x}`);
                            openMarkers.pop();
                        } else // something is still open and this isn’t a match -- might just be consequential error
                            if (char !== '’' // Closing single quote is also used as apostrophe in English
                                && (textType !== 'markdown' || char !== '>' || characterIndex > 4)) { // Markdown uses > or >> or > > or > > > for block indents so ignore these -- might just be consequential error
                                const excerpt = (characterIndex > excerptHalfLength ? '…' : '') + line.substring(characterIndex - excerptHalfLength, characterIndex + excerptHalfLengthPlus).replace(/ /g, '␣') + (characterIndex + excerptHalfLengthPlus < line.length ? '…' : '')
                                const details = `'${lastEntry.char}' opened on line ${lastEntry.n} character ${lastEntry.x + 1}`;
                                addNotice({ priority: 777, message: `Bad punctuation nesting: ${char} closing character doesn’t match`, details, lineNumber: n, characterIndex, excerpt, location: ourLocation });
                                // debugLog(`  ERROR 777: mismatched characters: ${details}`);
                            }
                    } else // Closed something unexpectedly without an opener
                        if (char !== '’' // Closing single quote is also used as apostrophe in English
                            && (textType !== 'markdown' || char !== '>')) { // Markdown uses > for block indents so ignore these
                            const excerpt = (characterIndex > excerptHalfLength ? '…' : '') + line.substring(characterIndex - excerptHalfLength, characterIndex + excerptHalfLengthPlus).replace(/ /g, '␣') + (characterIndex + excerptHalfLengthPlus < line.length ? '…' : '')
                            addNotice({ priority: 774, message: `Unexpected ${char} closing character (no matching opener)`, lineNumber: n, characterIndex, excerpt, location: ourLocation });
                            // debugLog(`  ERROR 774: closed with nothing open: ${char}`);
                        }
                }

            }
        } else {
            // This is a blank line
        }

        // lastLineContents = line;
    }
    //  At the end of the text -- check for left-over opening characters (unclosed)
    if (openMarkers.length) {
        const [{ char, n, x }] = openMarkers.slice(-1);
        const line = lines[n - 1];
        const excerpt = (x > excerptHalfLength ? '…' : '') + line.substring(x - excerptHalfLength, x + excerptHalfLengthPlus).replace(/ /g, '␣') + (x + excerptHalfLengthPlus < line.length ? '…' : '')
        const details = openMarkers.length > 1 ? `${openMarkers.length} unclosed set${openMarkers.length === 1 ? '' : 's'}` : null;
        addNotice({ priority: 768, message: `At end of text with unclosed ${char} opening character`, details, lineNumber: n, characterIndex: x, excerpt, location: ourLocation });
    }

    // TODO: Is this a duplicate of the above section about nesting?
    // Check matched pairs in the entire file
    for (const punctSet of OPEN_CLOSE_PUNCTUATION_PAIRS) {
        // Can’t check '‘’' coz they might be used as apostrophe
        const leftChar = punctSet[0], rightChar = punctSet[1];
        const leftCount = countOccurrencesInString(plainText, leftChar),
            rightCount = countOccurrencesInString(plainText, rightChar);
        if (leftCount !== rightCount
            && (rightChar !== '’' || leftCount > rightCount) // Closing single quote is also used as apostrophe in English
            && (textType !== 'markdown' || rightChar !== '>')) // markdown uses > as a block quote character
            // NOTE: These are lower priority than similar checks in a field
            //          since they occur only within the entire file
            addNotice({ priority: leftChar === '“' ? 162 : 462, message: `Mismatched ${leftChar}${rightChar} characters`, details: `left=${leftCount.toLocaleString()}, right=${rightCount.toLocaleString()}`, location: ourLocation });
    }

    if (!checkingOptions?.suppressNoticeDisablingFlag) {
        // functionLog(`checkPlainText: calling removeDisabledNotices(${cptResult.noticeList.length}) having ${JSON.stringify(checkingOptions)}`);
        cptResult.noticeList = removeDisabledNotices(cptResult.noticeList);
    }

    addSuccessMessage(`Checked all ${lines.length.toLocaleString()} line${lines.length === 1 ? '' : 's'}${ourLocation}.`);
    if (cptResult.noticeList.length)
        addSuccessMessage(`checkPlainText v${PLAIN_TEXT_VALIDATOR_VERSION_STRING} finished with ${cptResult.noticeList.length ? cptResult.noticeList.length.toLocaleString() : "zero"} notice${cptResult.noticeList.length === 1 ? '' : 's'}`);
    else
        addSuccessMessage(`No errors or warnings found by checkPlainText v${PLAIN_TEXT_VALIDATOR_VERSION_STRING}`)
    // debugLog(`  checkPlainText returning with ${result.successList.length.toLocaleString()} success(es), ${result.noticeList.length.toLocaleString()} notice(s).`);
    // debugLog("checkPlainText result is", JSON.stringify(result));
    return cptResult;
}
// end of checkPlainText function
