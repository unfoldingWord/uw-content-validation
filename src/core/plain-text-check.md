## Plain Text Check Sandbox

This function checks the given text for typical formatting errors.

It returns a list of success messages and a list of notice components. (There is always a priority number in the range 0..999 and the main message string, as well as other details to help locate the error as available.)

These raw notice components can then be filtered and/or sorted as required by the calling program, and then divided into a list of errors and a list of warnings or whatever as desired.

```js
// The control code in this box is editable for changing settingsâ€”
//    simply click inside here and add, change, or delete text as required.
// Note that (gray) lines starting with // are "comments", i.e., they are ignored by the software
//    so if you want to enable those lines, you must remove the // from the beginning of the line.

import { checkPlainText } from './plain-text-check';
import { RenderNumberedLines, RenderRawResults } from '../demos/RenderProcessedResults';

// Plain text samples
const textSG = `Short Good Plain Test

This is a paragraph.

Second level heading

Another paragraph.

  List item 1
  List item 2
`;
const textSB = `Short Ba,d Plain Test

This should be a  paragraph.

First level heading

Another  paragraph.

  List item 1
   List item 2
`;

// You can choose any of the above texts here
//  (to demonstrate differing results)
const username = 'unfoldingWord';
const chosenText = textSB;
const chosenTextName = 'textSB';
const checkingOptions = {};

const rawResults = checkPlainText(username, 'en', 'TN', 'raw', chosenTextName, chosenText, 'that was supplied', checkingOptions);
if (!rawResults.successList || !rawResults.successList.length)
  rawResults.successList = ["Done plain text checks"];

<>
<b>Source (normalized)</b>: <RenderNumberedLines text={chosenText} />
<RenderRawResults results={rawResults} />
</>
```
