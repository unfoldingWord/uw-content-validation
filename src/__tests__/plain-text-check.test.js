/* eslint-env jest */

import { checkPlainText } from '../core/plain-text-check';
import Path from "path";
import fs from 'fs-extra';

const optionalCheckingOptions = {
  originalLanguageRepoUsername: 'unfoldingWord',
  taRepoUsername: 'unfoldingWord',
  disableAllLinkFetchingFlag: true, // until we can solve localforage error: No available storage method found
  // The following flags have no meaning if the above is set to true
  disableLinkedTAArticlesCheckFlag: true,
  disableLinkedTWArticlesCheckFlag: true,
  disableLexiconLinkFetchingFlag: true,
  disableLinkedLexiconEntriesCheckFlag: true,
  getFile: params => {
    const { username, repository, path } = params;
    // console.log(`plain-text-check.test getFile(${username}, ${repository}, ${path})`)
    const filePath = Path.join('./src/__tests__/fixtures', username, repository, path);
    if (fs.existsSync(filePath)) {
      return fs.readFileSync(filePath).toString();
    }
    // eslint-disable-next-line no-throw-literal
    throw `plain-text-check.test getFile(): Could not find ${filePath}`;
  }
}

describe('checkPlainText() - ', () => {
  const username = 'unfoldingWord';
  const languageCode = 'en';
  const repoCode = 'LT';

  describe('short snippet tests - ', () => {
    it('should pass on smallest snippet', async () => {
      const plainText = `This is a short sentence that's all grammatically correct.
`;
      const rawResults = await checkPlainText(username, languageCode, repoCode, 'text', 'snippet', plainText, 'from test snippet', optionalCheckingOptions);
      console.log(`checkPlainText from plainText='${plainText}' got rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(0);
      expect(rawResults).toMatchSnapshot();
    });

    it('should fail on empty string', async () => {
      const rawResults = await checkPlainText(username, languageCode, repoCode, 'text', 'snippet', "", 'from test snippet', optionalCheckingOptions);
      console.log(`checkPlainText rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(1);
      expect(rawResults).toMatchSnapshot();
    });

    it('might fail on tabbed line', async () => {
      const plainText = "2:12\tgnn5\t\tfigs-parallelism\tשְׁלֵמָ֗ה\t1\tThis is a poetic expression that is very similar to the previous sentence. Alternate translation: “May Yahweh fully give to you everything that you deserve” (See: [[rc://*/ta/man/translate/figs-parallelism]], [Doublet](../figs-doublet/01.md))\n";
      const rawResults = await checkPlainText(username, languageCode, repoCode, 'text', 'snippet', plainText, 'from test snippet', optionalCheckingOptions);
      console.log(`checkPlainText from plainText='${plainText}' got rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(0);
      expect(rawResults).toMatchSnapshot();
    });

    it('should fail on separated digits', async () => {
      const rawResults = await checkPlainText(username, languageCode, repoCode, 'text', 'snippet', "Jesus fed 5, 000 people (or was it 4, 999?).\n", 'from test snippet', optionalCheckingOptions);
      console.log(`checkPlainText rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(3);
      expect(rawResults).toMatchSnapshot();
    });

    it('should fail on long digit string', async () => {
      const rawResults = await checkPlainText(username, languageCode, repoCode, 'text', 'snippet', "Jesus fed 5000 people (or was it 456789?).\n", 'from test snippet', optionalCheckingOptions);
      console.log(`checkPlainText rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(2);
      expect(rawResults).toMatchSnapshot();
    });

  });

});
