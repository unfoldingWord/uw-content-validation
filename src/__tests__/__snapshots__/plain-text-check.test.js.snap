// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkPlainText() -  short snippet tests -  might fail on tabbed line 1`] = `
Object {
  "noticeList": Array [],
  "successList": Array [
    "Checked all 2 lines from test snippet",
    "No errors or warnings found by checkPlainText v1.0.0",
  ],
}
`;

exports[`checkPlainText() -  short snippet tests -  should fail on empty string 1`] = `
Object {
  "noticeList": Array [
    Object {
      "characterIndex": -1,
      "excerpt": "",
      "location": " from test snippet",
      "message": "File ends without newline character",
      "priority": 538,
    },
  ],
  "successList": Array [
    "Checked all 1 line from test snippet",
    "checkPlainText v1.0.0 finished with 1 notice",
  ],
}
`;

exports[`checkPlainText() -  short snippet tests -  should fail on long digit string 1`] = `
Object {
  "noticeList": Array [
    Object {
      "excerpt": "5000",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Possible missing separator in digit string",
      "priority": 91,
    },
    Object {
      "excerpt": "456789",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Possible missing separator in digit string",
      "priority": 91,
    },
  ],
  "successList": Array [
    "Checked all 2 lines from test snippet",
    "checkPlainText v1.0.0 finished with 2 notices",
  ],
}
`;

exports[`checkPlainText() -  short snippet tests -  should fail on separated digits 1`] = `
Object {
  "noticeList": Array [
    Object {
      "characterIndex": 12,
      "excerpt": "…sus fed 5, 000 peopl…",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Unexpected leading zero",
      "priority": 92,
    },
    Object {
      "excerpt": "5, 000",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Found separated digits",
      "priority": 498,
    },
    Object {
      "excerpt": "4, 999",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Found possible separated digits",
      "priority": 198,
    },
  ],
  "successList": Array [
    "Checked all 2 lines from test snippet",
    "checkPlainText v1.0.0 finished with 3 notices",
  ],
}
`;

exports[`checkPlainText() -  short snippet tests -  should pass on smallest snippet 1`] = `
Object {
  "noticeList": Array [],
  "successList": Array [
    "Checked all 2 lines from test snippet",
    "No errors or warnings found by checkPlainText v1.0.0",
  ],
}
`;
